flow/machine.go

```diff
@@ -0,0 +1,121 @@
+package flow
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Machine describes the contract for the flow handling
+type Machine interface {
+	// Back move backward by step
+	Back(state *State) error
+	// Next move forward by step
+	Next(state *State) error
+	// ToStep Move to the step
+	ToStep(step int, state *State) error
+	// Success stop processing and call the final function
+	Success(state *State) error
```

Comment:
Not sure what value Success/Fail determination would add beyond general-purpose state abstraction. My intuition would be to simply process fails using a custom error type that would encompass all the information necessary for propagation.